### DISCUSSION TITLE
Bed levelleing mechanism & unsafe out pins

### DISCUSSION THREAD

**Discussion by xblax** [xblax/flashforge_ad5m_klipper_mod#6] [General]
Hi all,
@g992 @KaruroChori @jlodew0 @nitronov @gpo123 @ellaleks @AJolly @NokHook 

Here's a warning for you to not set the Level pins via Klipper, especially NOT use the Level H1, Level H2, Level H3 pins. They change the load cell calibration, I found this out the hard wary yesterday when I just toggled them to see what they do. I will hide them from Mainsail/Fluidd UI before the next release.

To re-calibrate my printer properly, I need your assistance: 
The bed leveling basically works like a kitchen scale. A load cell in the bed detects how hard the nozzle presses on the bed. The weight measurement is visible within Klipper/Mainsail as `[temperature_sensor weightValue]`.

![grafik](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/9802544/95037669-dc59-44d2-8463-838bb536934e)

To recalibrate my printer I need to now how it is calibrated when stock. Calibration works via the following mechanism:

Level H1 Pin, sets the zero point (no weight)
Level H2 Pin, sets the point where the sensor reads 500
Level H3 Pin, sets the point where the sensor reads 200

Can someone please put a weight on to the print bed so that the sensors reads near 500 and 200 and tell me the weight that was needed? Maybe it's 500g and 200g, but just a guess. Or any other sensor readouts as reference points would help too.

P.S.: I found this out via disassembling the flashforge software. There is a hidden maintenance menu to do the calibration.
![grafik](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/9802544/30b30031-a245-4db5-9340-23ceedbda7aa)

**Comment by KaruroChori** [comment]
Does it mean you are able to calibrate it back yourself? Otherwise I can perform this test for you.

**Comment by jlodew0** [comment]
@xblax i see in the calibration menu that the extruder is in the middle of build plate 110 110 and z-height 5. Does the move++ and - - change the z-height? Because I to me that would seem like you would need to apply Some pressure with the nozzle. I could place somerhing on the build plate but the size of the object would also influence the pressure Because of the weight distribution over a given surface.

**Comment by xblax** [comment]
@KaruroChori 
Yes, I need a reference point to do the calibration. I need to know how much weight I need to put on the bed for the sensor to read value 500, for example.

@jlodew0 
(110, 110) is the Endstop position, center is at (0,0). Yes, the move buttons change z-height, but calibration is not done via the nozzle. It should be without load and I think they put a reference weight on the bed. That behaves quite uniform, like a kitchen scale. So 500ml/500g water bottle would be interesting, what the weight sensor shows for you.

**Reply by santahx** [reply]
Hey i would Like to know how to callibrate it to be able to Print with an Glas bed installed.

**Comment by g992** [comment]
it is also most likely necessary to calibrate at a table temperature of 0 (room temperature), since the values depend very much on it
I didnt touch anything, 
I can try to set 500g something and tell u sensor value

**Comment by jlodew0** [comment]
I wonder if they are all (the printers) individually calibrated. And where these values are stored. The motherboards are for sale so I guess they are not stored on the onboard flash flash of the motherboard?

**Reply by xblax** [reply]
I think the calibration is stored in the load cell chip or maybe the sensor amplifier chip - so in the bed itself not on the motherboard. I have a generic micro scale and calibration works quite similar, via reference weights.

I guess they do the calibration individually after manufacturing.

**Comment by jlodew0** [comment]
My standard 'temperature' without any load gives me 720 degree for Weightvalue.
With 500gr in the middle it reads 140 degrees centigrade. Should I try to get it to 0?

**Reply by xblax** [reply]
So it decreases when you put weight on it? That's not what I expected.

**Reply by jlodew0** [reply]
yes, I got it to 0 degrees now with 620 grams of weight

**Comment by KaruroChori** [comment]
Assuming the bed is removed:
69g will lead to 480-500°C oscillating.
360g will lead to 200°C
550g will lead to 0°C

If you apply more load, the temperature shown will start raising again. So it is not a linear curve, but the internal calibration curve is likely a parabola.

**Reply by KaruroChori** [reply]
Please be mindful that the quality of my kitchen scale is not particularly high :D.
None of the calibrated scales I have access to will work with that much load, so I cannot provide anything better than that.

**Comment by xblax** [comment]
@jlodew0 @KaruroChori 
Thank you, that are good data points. Hopefully I can figure out appropriate calibration weights. Maybe it's also like the 'Tare' function of a scale, where they offset the 0-value.

Flashforge also seems to some conditional offset reset via level_clear and level H1 pins before leveling. Maybe that is to compensate changes in sensor readout due to heating the bed. I have to try understand that a bit better.

**Comment by jlodew0** [comment]
I'm pretty sure they have used resistive load cells (as oposed to capacitive ones), since a max31856 is used. (I used those adc's for k-probe temperature mesurements in some arduino projects in the past).

As for linearity the max datasheet states:
A lookup table (LUT) stores linearity correction data for several types of thermocouples (K, J, N, R, S, T, E, and B). Line frequency filtering of 50Hz and 60Hz is included, as is thermocouple fault detection. A SPI-compatible interface allows selection of thermocouple type and setup of the conversion and fault detection processes.

We could ask FF by email? I doubt that we will get an answer, but it doesn't harm to try.

**Comment by xblax** [comment]
@jlodew0 

> We could ask FF by email? I doubt that we will get an answer, but it doesn't harm to try.

Yes that's my plan B to ask how to perform the calibration, but they will wonder how I found that menu :D

Btw. the posted values seem to be more or less linear. The sensor value seems to decrease by 1 for each gram added to the bed. There is just a slight offset between your measurements.

I wouldn't bet on them really using a max31856. Maybe that is also a clone chip or just a similar chip with compatible interface.

**Comment by xblax** [comment]
Ok, so I think I figured this out more or less.

Pin H1, sets the zero point. It works like "Tare" on a scale.
Pin H3 adjusts the weight that equals a sensor reading of 500.
Pin H2, at least for me does have no meaningful effect. The scale seems to work linear, no matter what I try to set here.

So why does the sensor readout idle at a high value? 
I think that's because the Flashforge software always does a "Tare" before leveling, if the sensor value is off. Then, when the bed cools down it again shows a positive value (even though the weight is negative), similiar to "Tare" on a regular scale, just the negative sign removed.

So I used 500g to calibrate the weight.

With not other settings changed I get the following accuracy:

```
probe accuracy results: maximum -2.707500, minimum -2.752500, range 0.045000, average -2.725750, median -2.723750, standard deviation 0.014708
22:02
probe at 0.000,0.000 is z=-2.722500
22:02
probe at 0.000,0.000 is z=-2.712500
22:02
probe at 0.000,0.000 is z=-2.752500
22:02
probe at 0.000,0.000 is z=-2.725000
22:02
probe at 0.000,0.000 is z=-2.707500
22:02
probe at 0.000,0.000 is z=-2.747500
22:02
probe at 0.000,0.000 is z=-2.727500
22:02
probe at 0.000,0.000 is z=-2.707500
22:02
probe at 0.000,0.000 is z=-2.720000
22:02
probe at 0.000,0.000 is z=-2.735000
22:02
PROBE_ACCURACY at X:0.000 Y:0.000 Z:-0.742 (samples=10 retract=2.000 speed=5.0 lift_speed=5.0)
```
That range 0.045000 is too high. Klipper docs say it should be around 0.0125 or less for accurate first layers: https://www.klipper3d.org/Probe_Calibrate.html
The Sweetspot for me seems to be with 2mm/s probing speed.

```
22:05
probe accuracy results: maximum -2.627500, minimum -2.635000, range 0.007500, average -2.633750, median -2.635000, standard deviation 0.002305
22:05
probe at 0.000,0.000 is z=-2.627500
22:05
probe at 0.000,0.000 is z=-2.635000
22:05
probe at 0.000,0.000 is z=-2.635000
22:05
probe at 0.000,0.000 is z=-2.632500
22:05
probe at 0.000,0.000 is z=-2.632500
22:05
probe at 0.000,0.000 is z=-2.635000
22:05
probe at 0.000,0.000 is z=-2.635000
22:05
probe at 0.000,0.000 is z=-2.635000
22:05
probe at 0.000,0.000 is z=-2.635000
22:05
probe at 0.000,0.000 is z=-2.635000
22:05
PROBE_ACCURACY at X:0.000 Y:0.000 Z:-0.637 (samples=10 retract=2.000 speed=2.0 lift_speed=5.0)
22:05
PROBE_ACCURACY PROBE_SPEED=2 PROBE_RETRACT=2
```

What results do you get?

**Reply by AJolly** [reply]
PROBE_ACCURACY at X:0.000 Y:0.000 Z:3.240 (samples=40 retract=2.000 speed=2.0 lift_speed=5.0)
probe accuracy results: maximum -1.647500, minimum -1.690000, range 0.042500, average -1.673375, median -1.673750, standard deviation 0.008433


Yours is much better than mine :/

**Reply by xblax** [reply]
Would be funny if I get better results because I calibrated the load the cell after messing with the pins. I guess you did tare the load cell before probing? To check how your load cell is calibrated you can tare it an put a weight on the bed (i.e. 500g). I calibrated the load cell so that 500g weight give a sensor readout of 500. I think that in the end determines the probing force.

**Reply by AJolly** [reply]
ps. I suspect H2 sets the weight that's required to trigger the probe as closed.

**Comment by xblax** [comment]
Here is the sequence of calibration / self-test commands that the Flashforge software does before each probing / leveling:

```
{"id":566,"method":"gcode/script","params":{"script":"QUERY_BUTTON button=check_level_pin"}}.< 1970/01/01 00:15:19.928316  length=70 from=395599 to=395668
{"id":567,"method":"gcode/script","params":{"script":"SET_PIN PIN=level_h1 VALUE=0"}}.< 1970/01/01 00:15:19.937526  length=23 from=395692 to=395714
{"id":569,"method":"gcode/script","params":{"script":"SET_PIN PIN=level_h1 VALUE=1"}}.< 1970/01/01 00:15:20.444345  length=23 from=396456 to=396478
{"id":570,"method":"gcode/script","params":{"script":"SET_PIN PIN=level_h1 VALUE=0"}}.< 1970/01/01 00:15:20.951810  length=23 from=396574 to=396596
{"id":572,"method":"gcode/script","params":{"script":"SET_PIN PIN=level_h1 VALUE=1"}}.< 1970/01/01 00:15:21.458382  length=23 from=397331 to=397353
{"id":573,"method":"gcode/script","params":{"script":"QUERY_BUTTON button=check_level_pin"}}.< 1970/01/01 00:15:21.968302  length=69 from=397354 to=397422
{"id":574,"method":"gcode/script","params":{"script":"SET_PIN PIN=level_clear VALUE=0"}}.< 1970/01/01 00:15:21.977766  length=23 from=397446 to=397468
{"id":575,"method":"gcode/script","params":{"script":"SET_PIN PIN=level_clear VALUE=1"}}.< 1970/01/01 00:15:21.983983  length=23 from=397469 to=397491
```
The Level H1 pin, sets the load cell tare. That is important to adjust the load cell for the set bed temperature.

H1 is safe to use and we need that as well. I believe the other commands are to implement a safety mechanism, to not attempt bed leveling if the load cell is not connected:

- Button check_level_pin simply acts as an input pin because Klipper has no general purpose input pins otherwise
- Whenever, tare is set via H1, check_level_pin is activated (button becomes "pressed")
- level_clear pin, resets check_level_pin (button becomes "released")

So the sequence works this way:

1. Make sure that check_level_pin is not set
2. Set load cell tare via H1
3. Make sure that check_level_pin is set, raise error if not
4. Reset check_level_pin via level_clear pin

I will implement this as a Macro and overwrite `BED_MESH_CALIBRATE` to set the load cell tare before leveling.

**Comment by AJolly** [comment]
Any idea how to access the calibration menu on the stock fw? I was trying to get a thick G10 plate to work.

**Reply by xblax** [reply]
You mean that hidden menu? You don't need it. I could only access that menu by patching the software. So the 'License' button now opens that menu for me.

The software does automatic Tare of the load cell before leveling. It's fine to just put on a heavier plate. But how thick is that plate? If it's too thick you may risk crashing the head before leveling.

**Comment by AJolly** [comment]
Oh I definitely had it crashed into the plate with an ultra thick g10 plate.  (Had the same issues with my kobra 2 had to tweak a bunch of the clipper configs)

But even with the stock fw + plate, I did not have a great first layer: this was right out of the box after running abl.

![IMG20240222140658](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/91493/e55260f6-3e39-44b6-892e-bc36f9278da1)

![IMG20240222140717](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/91493/497be4a9-0956-46c1-9c44-8b17fa1c9ec0)
![IMG20240222140721](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/91493/d37e3cf8-9faf-49c0-9aab-2c204dd0e73b)
![IMG20240222140732](https://github.com/xblax/flashforge_adm5_klipper_mod/assets/91493/961180cc-b6af-41a3-8cf3-b417c9dda1d2)


PS how are you getting the sequence of commands from flash forge? Is this  sniffing port 8899 traffic or are we dumping the software on the printer itself?  (I used to be a professional reverse engineer but that was 15 years ago)

**Reply by xblax** [reply]
I think stock software bed leveling isn't that great because of how they configured the probe in Klipper. It doesn't have very high accuracy because the probe speed is too fast.

For thick plates, you have to change the config and macros in multiple places at the moment, at least with the mod klipper you can do it :) By default it's set up for PEI sheets. The head moves to 2mm before probing, and the bed is approx at -2mm. Not sure how much that varies between printers.

> PS how are you getting the sequence of commands from flash forge? Is this sniffing port 8899 traffic or are we dumping the software on the printer itself? (I used to be a professional reverse engineer but that was 15 years ago)

Socat sniffing on Unix socket /tmp/uds directly on the printer. That captures communication between Klipper and Flashforge Software. The protocol format is just JSON-RPC, so it's human readable.

**Comment by AJolly** [comment]
Are we sure about the 500g weight?
Here's my readings. The 50g weight is a calibrated reference weight.  All the other readings were done on a  reasonably precise scale.  I tared the 5m first.
![image](https://github.com/xblax/flashforge_ad5m_klipper_mod/assets/91493/82847f26-170c-4afb-895d-e1f2c4a07c2a)

Just took a few more readings.  The probe triggers when the load cell changes from reading 198 to 216.
215.1g was when the 5m would toggle back and forth between reading 198 and reading 216.

**Comment by AJolly** [comment]
I got the official instructions on calibrating the weight 

步骤1，进入设备信息界面，长按“设备信息”5s以上。机器会自动跳出如下界面，点击右上角“x”关闭这个界面。
注意，请勿点击这个界面上的任何按钮，否则可能更改机器配置，直接关闭界面即可。
Step 1: Enter the machine info interface and press "machine info" for more than 5 seconds. The machine will automatically pop up the following interface, click "x" in the upper right corner to close this interface.

Attention：please do not click any button on this interface, otherwise the machine configuration may be changed. Simply close the interface.

步骤2，回到设备信息界面，点击“调平开关”跳出调整界面。点击“H1 Point”，喷头和平台会走到待测试位置。
Step 2： return to the machine info interface and click on the "level sensor" to pop up the adjustment interface. Click on "H1 Point" and the extruder and platform will move to the position to be tested.

步骤3，点击H1后，确保重量值为0。
Step 3, after clicking H1, ensure that the weight value is 0.

步骤4，在打印平台上放置500g砝码，或500g重的物品（瓶装水等）。点击H2，查看重量数值，待数值稳定后，重量值为500g±30g则合格。
Step 4: Place a 500g weight or 500g  item (bottled water, etc.) on the printing  platform. Click H2 to view the weight value. After the value stabilizes, a weight value of 500g ± 30g is considered qualified.


注意：点击H1后，放置500g砝码或其他物品，如重量显示2048，是正常的，点击H2后能正常显示砝码重量即可。
Note: After clicking H1, place a 500g weight or other item. If the weight displays 2048, it is normal. After clicking H2, the weight can be displayed normally.

步骤5，点击H3后，点击下方关闭按钮，关闭当前界面。
Step 5: after clicking H3, click the close button below to close the current interface.

**Reply by xblax** [reply]
Cool! Did they just send you this via support? Not sure what they mean with 2048, afaik that's the max value that the sensor can read. Still sounds to me like it's calibrated to 500g via H2. Your stock measurements also support that.
